resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -467,20 +526,97 @@ private boolean handleRedirect(
                         continue;
                     }
 
-                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
-                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                    final InetSocketAddress socketAddress = parent.newRedirectServerAddress(resolved);
+                    nameServers.add(socketAddress);
+
+                    addNameServerToCache(authoritativeNameServer, socketAddress, r.timeToLive());
+                }
+
+                // Process all unresolved nameservers as well.
+                for (;;) {
+                    AuthoritativeNameServer authoritativeNameServer = serverNames.handleNext();
+
+                    if (authoritativeNameServer == null) {
+                        break;
+                    }
+                    // These will be resolved on the fly if needed.
```

Comment:
what if these entries exist in the cache already? The default comparator in `uncachedRedirectDnsServerStream` may prefer resolved addresses when compared to unresolved addresses. However it is possible that we are altering the order in which results are returned in the `RESPONSE` record because the server didn't directly provide the IP address, but we may already know the resolved address via the cache. I think we should cache this intermediate cname/redirect element, but also try to satisfy the IP address(es) via the cache.

Reply:
@Scottmitch I think I dont understand your comment ... Can you explain more what problems you see ? Is it because the order in the cache may be different then the order that is used by `uncachedRedirectDnsServerStream` ? Are you saying we should use the same order and even only put in the cache what is returned by  `uncachedRedirectDnsServerStream `?

Reply:
@Scottmitch @trustin @carl-mastrangelo WDYT about the "ordering" question here ?

Reply:
the comment was meant to highlight that we get results back from the server (e.g. Authority: A, B, C; Additional: B:IPx) and by default we reorder them according to which ones have the IPs specified in the additional section (B:IPx, A, C) ... however even if the server doesn't include a corresponding Additional entry, we may know the IP from the cache and can preserve the original ordering from the server (e.g. A:IPy, B:IPx, C)

IIRC the server responds in precedence order, so we should make a best effort to preserve this. We may want to prioritize name servers that we have a resolved address for (from cache or additional section), but potentially the unresolved DNS server will be more localized for the user and end up providing better performance in the long run?